src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+                case 2: return Components;
+                case 3: return CloseBracketToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```a-z``` node in a character class.
+    /// </summary>
+    internal sealed class RegexCharacterClassRangeNode : RegexPrimaryExpressionNode
+    {
+        public RegexCharacterClassRangeNode(
+            RegexExpressionNode left, RegexToken minusToken, RegexExpressionNode right)
+            : base(RegexKind.CharacterClassRange)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(minusToken.Kind == RegexKind.MinusToken);
+            Debug.Assert(right != null);
+            Left = left;
+            MinusToken = minusToken;
+            Right = right;
+        }
+
+        public RegexExpressionNode Left { get; }
+        public RegexToken MinusToken { get; }
+        public RegexExpressionNode Right { get; }
+
+        internal override int ChildCount => 3;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return Left;
+                case 1: return MinusToken;
+                case 2: return Right;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```-[f-m]``` in a pattern like ```[a-z-[f-m]]```.  A subtraction must come last in a 
+    /// character class, and removes some range of chars from the claracter class built up
+    /// so far.
+    /// </summary>
+    internal sealed class RegexCharacterClassSubtractionNode : RegexPrimaryExpressionNode
+    {
+        public RegexCharacterClassSubtractionNode(
+            RegexToken minusToken, RegexBaseCharacterClassNode characterClass)
+            : base(RegexKind.CharacterClassSubtraction)
+        {
+            Debug.Assert(minusToken.Kind == RegexKind.MinusToken);
+            Debug.Assert(characterClass != null);
+            MinusToken = minusToken;
+            CharacterClass = characterClass;
+        }
+
+        public RegexToken MinusToken { get; }
+        public RegexBaseCharacterClassNode CharacterClass { get; }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return MinusToken;
+                case 1: return CharacterClass;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// Represents a ```[:...:]``` node in a character class.  Note: the .net regex parser
```

Comment:
Not related to this PR: I think it makes sense to recognize backquotes and triple-backquotes in xml doc comments.
Any idea where I should look to add that?

![image](https://user-images.githubusercontent.com/12466233/38170603-f11338f6-353e-11e8-8e84-7c7b7d33eba3.png)
Filed https://github.com/dotnet/roslyn/issues/25867